### ISSUE TITLE
Hundreds or thousands of open files when sending traces to Stackdriver Trace in gRPC>=1.12.0 in Python

### ISSUE THREAD

**Issue by davewalkexpel** [grpc/grpc#17379]
### What version of gRPC and what language are you using?
Experiencing the problem with >1.12.0 (specifically experienced with 1.12.0, 1.15.0 and 1.16.1)
 
 
### What operating system (Linux, Windows, …) and version?
CentOS 7
 
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
 Python 3.5.5 (default, Feb  6 2018, 10:57:32)
[GCC 4.8.5 20150623 (Red Hat 4.8.5-16)] on linux
 
### What did you do?
I ran my application that is written to send traces to Google's Stackdriver Trace. It's using the [opencensus library](https://github.com/census-instrumentation/opencensus-python) (I'm pinned to 0.1.8), which uses the [google-cloud-trace](https://github.com/googleapis/google-cloud-python/tree/master/trace) package (pinned to >=0.17.0, <0.20) to actually forward the traces to Stackdriver via its gRPC service.

### What did you expect to see?
Traces in Stackdriver Trace, not thousands of open files
 
### What did you see instead?
Traces in Stackdriver Trace and open files, which point to a Google server. It looks like this:

```bash
python3.5 23896 root 1484u  IPv4 1537584      0t0  TCP 5b493ea568a2:56516->iad23s59-in-f10.1e100.net:443 (ESTABLISHED)
python3.5 23896 root 1485u  IPv4 1535805      0t0  TCP 5b493ea568a2:56520->iad23s59-in-f10.1e100.net:443 (ESTABLISHED)
python3.5 23896 root 1486u  IPv4 1538364      0t0  TCP 5b493ea568a2:56524->iad23s59-in-f10.1e100.net:443 (ESTABLISHED)
python3.5 23896 root 1487u  IPv4 1537592      0t0  TCP 5b493ea568a2:56528->iad23s59-in-f10.1e100.net:443 (ESTABLISHED)
python3.5 23896 root 1488u  IPv4 1538390      0t0  TCP 5b493ea568a2:56544->iad23s59-in-f10.1e100.net:443 (ESTABLISHED)
python3.5 23896 root 1489u  IPv4 1537628      0t0  TCP 5b493ea568a2:56546->iad23s59-in-f10.1e100.net:443 (ESTABLISHED)
python3.5 23896 root 1490u  IPv4 1538392      0t0  TCP 5b493ea568a2:56548->iad23s59-in-f10.1e100.net:443 (ESTABLISHED)
python3.5 23896 root 1491u  IPv4 1537634      0t0  TCP 5b493ea568a2:56550->iad23s59-in-f10.1e100.net:443 (ESTABLISHED)
python3.5 23896 root 1492u  IPv4 1538395      0t0  TCP 5b493ea568a2:56552->iad23s59-in-f10.1e100.net:443 (ESTABLISHED)
```

```bash
# lsof -i -P | grep e100 | grep ESTABLISHED | wc -l
1342
```

With `GRPC_VERBOSITY=DEBUG` and `GRPC_TRACE=call_error` set I get a lot of:

```bash
I1203 18:38:32.470243000   23908 subchannel.cc:656]          New connected subchannel at 0x7fcc0a805680 for subchannel 0x7fcb645ca030
D1203 18:38:32.660748300   23908 call.cc:700]                set_final_status CLI
D1203 18:38:32.660907100   23908 call.cc:701]                "No Error"
D1203 18:38:32.670058300   23908 dns_resolver.cc:279]        Start resolving.
```

over and over again.

And if I `watch` this it'll grow by about a dozen per second

I did some research on this and seems to have started with [1.12.0](https://github.com/grpc/grpc/releases/tag/v1.12.0) in which `channel.close()` is recommended. The google-cloud-trace and opencensus libraries don't do this, but I figure I'll patch opencensus in my docker container that I'm testing with to see if it helps. I essentially added a `self.client.channel.close()` after [this line](https://github.com/census-instrumentation/opencensus-python/blob/master/opencensus/trace/exporters/stackdriver_exporter.py#L224). I run my app again and the count of "ESTABLISHED" open files grows, but at a much slower rate now, about two every second. And it maxes out at about 386 files every time. When it starts to max out, the count of "CLOSE_WAIT" connections starts to increase at about the rate of one per second. The entire time the gRPC logs are spitting out "New connected subchannel at x for subchannel y" over and over again.

I'm not saying that this is a problem with the gRPC library, perhaps the tracing code is doing something funky that I'm not understanding, but I thought I would start here to see if any of this means anything to you. As far as I know the Google' libraries don't call close on channels; this has come up before: ([#5523](https://github.com/googleapis/google-cloud-python/issues/5523) and [#5570](https://github.com/googleapis/google-cloud-python/issues/5570). The general consensus from those issues is that the Google library isn't going to close the channel. Fine, so I did it myself, but I still have hundreds of open connections. When I do close the channel, is it expected for me to still have so many open connections? Any help would be appreciated!

**Comment by davewalkexpel** [comment]
Oh good call. My tracer now is created like this, copy/pasted from the link you provided:

```python
from opencensus.trace.exporters import stackdriver_exporter
from opencensus.trace import tracer as tracer_module
from opencensus.trace.exporters.transports.background_thread import BackgroundThreadTransport
exporter = stackdriver_exporter.StackdriverExporter(
    project_id='my-project', transport=BackgroundThreadTransport)
tracer = tracer_module.Tracer(exporter=exporter)
```

The problem remains, unfortunately. Let me know if I can provide anything else.

**Comment by ericgribkoff** [comment]
@davewalkexpel I didn't get a chance to look at this until today, and apparently I picked a good time. The latest opencensus release came out today and includes this commit https://github.com/census-instrumentation/opencensus-python/commit/d0fbdb99cd22110d2a1f1e1a9eb0a1b642b87ec0, which bumps google-cloud-trace to 0.20.1 to "fix a memory leak". I didn't dig into specifics of what was leaking, but from the log output it looks like each trace was creating its own gRPC channel, as you observed and reported here.

Bumping to opencensus==0.1.10 and google-cloud-trace==0.20.1, a single gRPC channel is correctly reused for all traces. 

I used the following code to test that a channel per trace was incorrectly recreated per trace using google-cloud-trace < 0.20.1, and now reusing a single gRPC channel correctly with google-cloud-trace 0.20.1. (Setting `transport=BackgroundThreadTransport` did not affect this behavior either way)

```
import google.auth
from opencensus.trace.exporters import stackdriver_exporter as stackdriver
from opencensus.trace import tracer as tracer_module

try:
  _, project_id = google.auth.default()
except google.auth.exceptions.DefaultCredentialsError:
  raise ValueError("Couldn't find Google Cloud credentials, set the project ID with 'gcloud set project'")

exporter = stackdriver.StackdriverExporter(project_id=project_id)
tracer = tracer_module.Tracer(exporter=exporter)

for i in range(10000):
  with tracer.span(name='span name') as span:
    print("doing work!")
```

I'm going to mark this closed, as it appears the problem in the upstream use of gRPC has been corrected. @davewalkexpel  Let me know if there's any thing I've missed here and I can re-open. Thanks.
